### PR TITLE
"#14 Fix missing link from last up block to Conv2D"

### PR DESCRIPTION
**Pull request by bguisard**
Includes 'x_up' on the upsampling dense blocks allows the 'include_top'
layer to get the full feature map at the end of the upsampling branch.

This commit also changes the 'concatenate' inside __dense_block as the
previous method was failing to run on TensorFlow backend.

This was tested using Keras 2.0.2 and TensorFlow 1.1.0

I was having issues when running the concatenate with your list method and my code works on tf, but I didn't test it with th backend.

Also I double checked that the dimensions make sense by comparing with the Lasagne implementation from Simon Jégou. The only thing that is still weird is that this has fewer connections than the table on their paper suggests. I trained it to find vehicles on a video stream and got very good results, but it's possible that there are still something missing.

### CODE REVIEW

File: densenet.py [hunk 1]
```diff
@@ -613,14 +614,14 @@ def __create_fcn_dense_net(nb_classes, img_input, include_top, nb_dense_block=5,
...
 
         # Dont allow the feature map size to grow in upsampling dense blocks
-        _, nb_filter, concat_list = __dense_block(x, nb_layers[nb_dense_block + block_idx + 1], nb_filter=growth_rate,
-                                                  growth_rate=growth_rate, dropout_rate=dropout_rate,
-                                                  weight_decay=weight_decay,
-                                                  return_concat_list=True, grow_nb_filters=False)
+        x_up, nb_filter, concat_list = __dense_block(x, nb_layers[nb_dense_block + block_idx + 1], nb_filter=growth_rate,
+                                                     growth_rate=growth_rate, dropout_rate=dropout_rate,
+                                                     weight_decay=weight_decay,
+                                                     return_concat_list=True, grow_nb_filters=False)
```
Comment (titu1994) [inline densenet.py]: Over here, add the line x = x_up

File: densenet.py [hunk 2]
```diff
@@ -613,14 +614,14 @@ def __create_fcn_dense_net(nb_classes, img_input, include_top, nb_dense_block=5,
...
 
     if include_top:
         x = Conv2D(nb_classes, (1, 1), activation='linear', padding='same', kernel_regularizer=l2(weight_decay),
-                   use_bias=False)(x)
+                   use_bias=False)(x_up)
```
Comment (titu1994) [inline densenet.py:624]: revert this to x

**Review by titu1994 — CHANGES_REQUESTED** [review]
Can you make the two changes, cause the return will be eroneous, if it return x, but x is not updated with x_up in the loop (basically half the graph will not be there if include_top is false).

Another thing is can you upgrade to Keras 2.0.4, and create a densenet-fcn 56 and 101 and see if the number of parameters is close to the number in the original paper? To do so, just build the model and then use model.summary() to see the final number of parameters. 

It should be close to the original paper, with a maximum of 5% margin of difference.

File: densenet.py (hunk 2 above)
Comment (bguisard) [reply densenet.py:624]: I think it's best if we leave this like they are, otherwise you will feed the full feature map into the upsampling dense block.

In my opinion the best way to address this is by adding:
x = x_up
inside the else statement on row 629

Comment (titu1994) [reply densenet.py:624]: Hmm I see. Ok that makes sense.

Comment (titu1994) [reply densenet.py:624]: Wait inside the else statement?? I didnt mean that. 

The return on line 635 will return x from line 614 inside the loop if we do not add x = x_up in the line that I mentioned in the above commit. There is no need for x = x_up in the line 629.

Comment (bguisard) [reply densenet.py:624]: Cool. That's done.

On the parameters I am seeing 1,127,952 parameters for FC56 instead of 1.5M and 8,167,504 for FC103 instead of 9.4M mentioned on the paper. The weird thing is that the ratio of missing connections is not growing in the same proportion.

Comment (titu1994) [reply densenet.py:624]: Hmm what are the parameters you are using for the model building? As I recall, the number of parameters was correct to about 5 % difference before.

Comment (titu1994) [reply densenet.py:624]: No I just checked again, seems my code was off by a lot as well. Yours is closer to theirs. I am getting just 8 M params for 103, whereas yours is 8.1 M. Still, I wonder where the remaining 1.3 M params are from..

Comment (bguisard) [reply densenet.py:624]: image_dim = (224, 224, 3)
FC56 = densenet.DenseNetFCN(input_shape=image_dim, growth_rate=12, dropout_rate=0.2,
                            init_conv_filters=24, activation='sigmoid')

layers_per_block = [4, 5, 7, 10, 12, 15]
tiramisu103 = densenet.DenseNetFCN(input_shape=image_dim, 
                                   nb_layers_per_block=layers_per_block,
                                   dropout_rate=0.2, activation='sigmoid')

**Review by titu1994 — APPROVED** [review]

File: densenet.py (hunk 2 above)
Comment (titu1994) [reply densenet.py:624]: Yup, yours is correct. I am still wondering where the remaining 1.3 M parameters are from, but yours is better.

Comment (bguisard) [reply densenet.py:624]: I am going to spend some more time investigating if we are still missing something, but the code I sent you is doing a very good job in detecting vehicles. You can check my implementation on https://github.com/bguisard/CarND-Vehicle-Detection if you are interested.

Comment (titu1994) [reply densenet.py:624]: Yup, will look into that. Thanks a lot for the fix. I will try this model out on one of my projects, because it failed to train properly before.

**Comment by titu1994** [comment]
Lol. I got the "bug". The default upsampling technique is Subpixel Convolutions. However, if you switch upsample_type="deconv", you will get exactly 9.4 million parameters now. 

Sorry for the confusion, and I will push the change now.

**Comment by bguisard** [comment]
Nice! I completely forgot about the fact we are using a different upsampling method. Great to see that it's all working properly.

Good luck with your project.
